embassy-nrf/src/lib.rs

```diff
@@ -549,18 +548,90 @@ pub fn init(config: config::Config) -> Peripherals {
                 }
             }
 
+            #[cfg(feature = "_nrf54l")]
+            {
+                use crate::pac::tampc::regs::DomainDbgenCtrl;
+                use crate::pac::tampc::regs::NidenCtrl;
+                use crate::pac::tampc::regs::SpidenCtrl;
+                use crate::pac::tampc::regs::SpnidenCtrl;
+                use crate::pac::tampc::vals::DomainDbgenCtrlValue;
+
+                // UICR cannot be written here, because it can only be written once after an erase all.
+                // The erase all value means that debug access is allowed if permitted by the firmware.
+
+                // Write to TAMPC to permit debug access
+                //
+                // See https://docs.nordicsemi.com/bundle/ps_nrf54L15/page/debug.html#ariaid-title6
+
+                let p = pac::TAMPC;
+
+                // Unlock dbgen
+                let mut writeprotect_clear = DomainDbgenCtrl::default();
+                writeprotect_clear.set_key(crate::pac::tampc::vals::DomainDbgenCtrlKey::KEY);
+                writeprotect_clear.set_writeprotection(crate::pac::tampc::vals::DomainDbgenCtrlWriteprotection::CLEAR);
+
+                let mut debug_enable = DomainDbgenCtrl::default();
+                debug_enable.set_key(crate::pac::tampc::vals::DomainDbgenCtrlKey::KEY);
+                debug_enable.set_value(DomainDbgenCtrlValue::HIGH);
+
+                p.protect().domain(0).dbgen().ctrl().write_value(writeprotect_clear);
+                p.protect().domain(0).dbgen().ctrl().write_value(debug_enable);
+
+                // Unlock niden
+                //
+                // Reuse values from dbgen disable
+                p.protect()
```

Comment:
this is quite verbose. I see you did it this way to reuse the values across dbgen, niden, spiden but it ends up being more verbose and nothing enforces the bits for these are at the same position. i'd suggest using the plain old closure `write` instead:

```rust
use crate::pac::tampc::vals;

p.protect().domain(0).niden().ctrl().write(|w| {
    w.set_key(vals::DomainDbgenCtrlKey::KEY);
    w.set_writeprotection(vals::DomainDbgenCtrlWriteprotection::CLEAR);
});
```